subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -299,1706 +417,5209 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+		} else {
+			control = net_buf_pull_le16(buf);
+			req_seq = BT_L2CAP_I_FRAME_ENH_CONTROL_GET_REQ_SEQ(control);
+			f = BT_L2CAP_I_FRAME_ENH_CONTROL_GET_F(control);
+			sar = BT_L2CAP_I_FRAME_ENH_CONTROL_GET_SAR(control);
+			tx_seq = BT_L2CAP_I_FRAME_ENH_CONTROL_GET_TX_SEQ(control);
+		}
 
-	if (BR_CHAN(chan)->state != BT_L2CAP_CONNECTING) {
-		return;
+		err = bt_l2cap_br_update_f(br_chan, f);
+		if (err) {
+			/* Invalid F bit. Ignore the I-frame. */
+			return;
+		}
+	} else {
+		control = net_buf_pull_le16(buf);
+		req_seq = BT_L2CAP_I_FRAME_STD_CONTROL_GET_REQ_SEQ(control);
+		sar = BT_L2CAP_I_FRAME_STD_CONTROL_GET_SAR(control);
+		r = BT_L2CAP_I_FRAME_STD_CONTROL_GET_R(control);
+		tx_seq = BT_L2CAP_I_FRAME_STD_CONTROL_GET_TX_SEQ(control);
+		bt_l2cap_br_update_r(br_chan, r);
 	}
 
-	LOG_DBG("chan %p status 0x%02x encr 0x%02x", chan, status, chan->conn->encrypt);
+	if (br_chan->rx.mode == BT_L2CAP_BR_LINK_MODE_STREAM) {
+		bt_l2cap_br_update_expected_tx_seq(br_chan, tx_seq + 1);
+		/* Missing report */
+		goto valid_frame;
+	}
 
-	if (status) {
-		/*
-		 * Security procedure status is non-zero so respond with
-		 * security violation only as channel acceptor.
-		 */
-		l2cap_br_conn_req_reply(chan, BT_L2CAP_BR_ERR_SEC_BLOCK);
+	/* The first valid I-frame received after an REJ was sent, with a TxSeq of the
+	 * received I-frame equal to ReqSeq of the REJ, shall clear the REJ Exception
+	 * condition.
+	 */
+	if (br_chan->req_seq == tx_seq) {
+		LOG_DBG("Seq %zu equals to req_seq, clears the REJ exception", tx_seq);
+		atomic_clear_bit(br_chan->flags, L2CAP_FLAG_SEND_FRAME_REJ);
+	}
 
-		/* Release channel allocated to outgoing connection request */
-		if (atomic_test_bit(BR_CHAN(chan)->flags,
-				    L2CAP_FLAG_CONN_PENDING)) {
-			l2cap_br_chan_cleanup(chan);
+	if (br_chan->expected_tx_seq == tx_seq) {
+		/* Valid TX seq received */
+		LOG_DBG("Valid information received seq %zu", tx_seq);
+		expected_tx_seq = true;
+	} else {
+		if (bt_l2cap_br_check_tx_seq_duplicated(br_chan, tx_seq)) {
+			/*
+			 * The Information field shall be discarded since it has already been
+			 * received.
+			 */
+			discard = true;
+			LOG_DBG("Duplicated Information received %zu", tx_seq);
+		} else if (bt_l2cap_br_check_tx_seq_out_of_sequence(br_chan, tx_seq)) {
+			/* The missing I-frame(s) are considered lost and ExpectedTXSeq is
+			 * set equal to TxSeq+1.
+			 */
+			LOG_DBG("Missing I-frame detected %zu", tx_seq);
+			if (br_chan->rx.mode == BT_L2CAP_BR_LINK_MODE_FC) {
+				expected_tx_seq = true;
+				/* TODO: Missing report */
+			} else {
+				/* A REJ exception is triggered, and an REJ frame with
+				 * ReqSeq equal to ExpectedTxSeq shall be sent to initiate
+				 * recovery. The received I-frame shall be discarded.
+				 */
+				discard = true;
+				LOG_DBG("Set REJ Exception");
+				if (!atomic_test_and_set_bit(br_chan->flags,
+							     L2CAP_FLAG_SEND_FRAME_REJ)) {
+					LOG_DBG("Set REJ Exception received flag");
```

Comment:
Some really deep indentation here. It doesn't look like this might be solvable with the inverse conditions approach, so you might want to look at other kinds of restructuring, e.g. refactoring into helper functions.

Reply:
Yes. It could be moved into a dedicated function. Let me update it.

Reply:
Updated.